### PR TITLE
Decouple Painless AST Lambda generation from the grammar

### PR DESCRIPTION
**Pull request by jdconrad**
This is the first step in decoupling the Painless AST from the grammar. The Painless AST should be able to generate classes independently of how the AST is generated from a grammar. (If I were to build a Painless AST by hand in code this should be all that's necessary.) This change removes Lambda name generation from the ANTLR grammar tree walker. It also removes unnecessary node generation of new array function references from the tree walker as well.

### CODE REVIEW

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/2

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java [hunk 1]
```diff
@@ -1115,22 +1106,9 @@ public ANode visitClassfuncref(ClassfuncrefContext ctx) {
...
-            String name = nextLambda();
-            globals.addSyntheticMethod(new SFunction(new FunctionReserved(), location, arrayType, name,
-                Arrays.asList("int"), Arrays.asList("size"), Arrays.asList(code), true));
-
-            return new EFunctionRef(location(ctx), "this", name);
-        }
-
-        return new EFunctionRef(location(ctx), ctx.decltype().getText(), ctx.NEW().getText());
+        return ctx.decltype().LBRACE().isEmpty() == false ?
+            new ENewArrayFunctionRef(location(ctx), ctx.decltype().getText()) :
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java]: can we flip this to use positive logic? if empty, function ref, else array function ref

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArrayFunctionRef.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        }
+    }
+
+    @Override
+    void write(MethodWriter writer, Globals globals) {
+        if (ref != null) {
+            writer.writeDebugInfo(location);
+            writer.invokeLambdaCall(ref);
+        } else {
+            // TODO: don't do this: its just to cutover :)
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArrayFunctionRef.java]: can we have a more descriptive comment?

**Review by rjernst — APPROVED** [review]
LGTM

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArrayFunctionRef.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ENewArrayFunctionRef.java]: Done.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review! Will commit once CI passes.

**Pull request merged**